neural_modelling/src/neuron/plasticity/stdp/synapse_dynamics_stdp_mad_neuromodulated_impl.c

```diff
@@ -0,0 +1,455 @@
+        post_event_history_t *post_event_history) {
+
+    use(&trace);
+
+    // Calculate EXP components in JK's weight update equation
+    int16_t decay_eligibility_trace = DECAY_LOOKUP_TAU_C(
+        time - last_non_dopamine_spike_time);
+    int16_t decay_dopamine_trace = DECAY_LOOKUP_TAU_D(
+        time - post_event_history -> last_dopamine_spike_time);
+    int16_t third_exp_component;
+    if (last_update_time == last_non_dopamine_spike_time)
+        third_exp_component = DECAY_LOOKUP_TAU_D(
+            last_non_dopamine_spike_time -
+            post_event_history -> last_dopamine_spike_time);
+    else
+        third_exp_component = DECAY_LOOKUP_TAU_C(
+             post_event_history -> last_dopamine_spike_time -
+             last_non_dopamine_spike_time);
+
+    // Evaluate weight function
+    uint32_t weight_change = STDP_FIXED_MUL_16X16(
+            STDP_FIXED_MUL_16X16(post_event_history -> last_neuromodulator_trace,
+                previous_state -> eligibility_trace),
+            STDP_FIXED_MUL_16X16(weight_update_constant_component,
+        STDP_FIXED_MUL_16X16(decay_eligibility_trace, decay_dopamine_trace)
+            - third_exp_component));
+
+    previous_state -> weight += weight_change;
+
+    if (previous_state -> weight & 0x8000)
+        previous_state -> weight = 0;
+    else
+        // Saturate weight
+        previous_state -> weight= MIN(weight_state.weight_region->max_weight,
+                                      MAX(previous_state -> weight,
+                                      weight_state.weight_region->min_weight));
+
+    // Update eligibility trace if this spike is non-dopamine spike
+    if (trace.dopamine == 0) {
+        // Decay eligibility trace
+        int32_t decayed_eligibility_trace = STDP_FIXED_MUL_16X16(
+            previous_state -> eligibility_trace, decay_eligibility_trace);
+
+        // Apply STDP
+        uint32_t time_since_last_pre = time - last_pre_time;
+        if (time_since_last_pre > 0) {
+            int32_t decayed_r1 = STDP_FIXED_MUL_16X16(
+                last_pre_trace, DECAY_LOOKUP_TAU_PLUS(time_since_last_pre));
+            decayed_eligibility_trace += decayed_r1;
+        }
+        previous_state -> eligibility_trace = decayed_eligibility_trace;
+    }
+}
+
+static inline void correlation_apply_pre_spike(
+        uint32_t time, pre_trace_t trace, uint32_t last_post_time,
+        post_trace_t last_post_trace, uint32_t last_non_dopamine_spike_time,
+        uint32_t last_update_time, plastic_synapse_t *previous_state,
+        post_event_history_t *post_event_history) {
+
+    use(&trace);
+    use(&last_post_trace);
+
+    // Calculate EXP components in JK's weight update equation
+    int16_t decay_eligibility_trace = DECAY_LOOKUP_TAU_C(
+        time - last_non_dopamine_spike_time);
+    int16_t decay_dopamine_trace = DECAY_LOOKUP_TAU_D(
+        time - post_event_history -> last_dopamine_spike_time);
+    int16_t third_exp_component;
```

Comment:
Again you shouldn't use 16-bit integers to STORE stuff that's going to live in registers or on the stack - ARM registers are 32-bit AND it results in bit masking operations whenever you access them